File: 2021-09/문자열내맘대로정렬.js
function solution(strings, n) {
    return strings
        .sort((a, b) => {
        if(a.charAt(n) < b.charAt(n)) return -1;
        if(a.charAt(n) > b.charAt(n)) return 1;
        if(a < b) return -1;
        if(a > b) return 1;
        return 0;
        //a.charAt(n) < b.charAt(n) ? -1 : a.charAt(n) > b.charAt(n) ? 1 : a < b ? -1 : a > b ? 1 : 0
    });
}

console.log(solution(["sun", "bed", "car"], 1)); // ["car", "bed", "sun"]

File: programmers/src/main/java/level1/문자열내맘대로정렬.java
package level1;

import java.util.Arrays;

public class 문자열내맘대로정렬 {

	public String[] solution(String[] strings, int n) {
		return Arrays.stream(strings)
			.map(st -> new RandomString(n, st))
			.sorted()
			.map(RandomString::toString)
			.toArray(String[]::new);
	}

	static class RandomString implements Comparable<RandomString> {

		private final int idx;
		private final String string;

		public RandomString(int idx, String string) {
			this.idx = idx;
			this.string = string;
		}

		@Override
		public int compareTo(RandomString o) {
			char a = getCompareChar();
			char b = o.getCompareChar();
			if (a == b) {
				return string.compareTo(o.toString());
			}
			return a > b ? 1 : -1;
		}

		@Override
		public String toString() {
			return string;
		}

		public char getCompareChar() {
			return string.charAt(idx);
		}
	}

}
